### ISSUE TITLE
missing example of training a gan "without requiring supervision of masks or points"

### ISSUE THREAD

**Issue by romain-rsr** [xingzhehe/GANSeg#4]
Hi,

We were able to run your gan on the celeba example but we are struggling to apply it on raw images (without any supervision of mask or keypoint). 

After careful reading of your works, we understand the main interest of your model is to generate keypoints then masks then segmented images from training on a dataset with only raw images, without any mask or keypoints in the training set. Can you please confirm this and provide us with a version of your code that can be applied on a folder containing only raw 256x256 images ? 

Many thanks again for your work and reactivity,
Romain

**Comment by xingzhehe** [comment]
Hi Romain,

Yes our main interest is obtaining the masks and keypoints by training only on raw images.

I am not sure I understand your request. Our model can be directly applied if you simply resize the 256x256 images to 128x128. 
If you want to re-train on 256x256 images, you can 
1) delete the hard coded architecture in line 149-225 in generator.py to allow it consume 256x256 images.. 
2) copy line 41 to line 42 in discriminator.py to allow it consume 256x256 images.

If you have further questions, don't hesitate to reach me.

Bests,
Xingzhe

**Comment by romain-rsr** [comment]
Hi,

I'm sorry i didn't go straight to the point : we run your model successfully on celeba but we failed to apply it on this toy example, where the raw images we train the model on are plain blue rectangles of various sizes and position on a plain grey background : 

![image1](https://user-images.githubusercontent.com/47384615/208165888-03878cdb-cd22-4345-90c4-92370adc0a0a.png)

The model succeeds to generate such shapes but fail into segmenting the blue rectangle from the background. 

- here are two example with 3 keypoints :

![image2](https://user-images.githubusercontent.com/47384615/208167179-662e2f90-9ff4-44b5-acff-045f3aa48e77.png)
![image3b](https://user-images.githubusercontent.com/47384615/208167175-8ba18153-226f-4555-a3cb-4ffe8572f113.png)
![image3](https://user-images.githubusercontent.com/47384615/208167178-c0b7056f-c2a2-4fe0-9c33-383fd407f444.png)
(the segmentation mask is a tiny point on this one)

![image4a](https://user-images.githubusercontent.com/47384615/208167913-dee51a75-b1e9-42cb-82df-bef7440fc1cc.png)
![image4b](https://user-images.githubusercontent.com/47384615/208167917-cb2c617e-48c8-4e45-b7ee-060ed9fa2dda.png)
![image5](https://user-images.githubusercontent.com/47384615/208167918-2d0832d1-0dd8-483d-8ec1-38a1975f783f.png)

- here is an other examples with 8 keypoints :

![image6](https://user-images.githubusercontent.com/47384615/208168073-a41d5329-4a6f-40a5-8ec6-8f3d7e0fa975.png)
![image7](https://user-images.githubusercontent.com/47384615/208168072-23f02850-79ef-4871-a1e3-ad304857c845.png)
![image8](https://user-images.githubusercontent.com/47384615/208168069-3d664d48-1d58-498c-b346-2a340e1e5d4e.png)

---------------------- more info

- the number of processing epochs for these failed experiments was similar (and actually higher) than the one for the succeeding runs on celeba.
- we used a the same batch size of 4 for all experiments (celeba and toy samples)
 
---------------------- why we asked about the preprocessing first

Since we produced the model input h5 file by applying the celeba processing file on our toy samples, our previous question aimed to verify that the preprocessing step was not in fault. Meanwhile, we actually wrote a generic preprocessing file on our own, that only requires a folder with raw images in it (without any segmentation information). It's available here if required : https://github.com/romain-rsr/colab/blob/main/uprocess.py

Bests,
Romain

**Comment by xingzhehe** [comment]
Hi Romain,

Thanks for this experiment! It is indeed very very interesting! It illustrates some insights I have never thought about.

I also tested by myself on GANSeg, and it didn't work. I think it is due to the (positional encoding or part embedding or background embedding) overfit this too simple dataset.

In parallel, I also tested on two different unsupervised keypoint detection methods, which use keypoint embeddings. They also fail although slightly slightly better than GANSeg, probably because they don't have background embeddings and positional encoding. I doubt if all methods using embeddings without additional care could lead to this problem to some extend.

Finally I tested on a method without using embeddings: https://xingzhehe.github.io/autolink/
and it finally works:
<img width="681" alt="image" src="https://user-images.githubusercontent.com/28001407/208178559-f952252c-071a-4f13-8101-c2e16a060271.png">
<img width="1142" alt="image" src="https://user-images.githubusercontent.com/28001407/208178823-0c9dbb79-ed9e-456a-8b25-1fc0ef7e815c.png">
Although it only detect keypoints and their linkages, masks can be easily extracted.

Bests,
Xingzhe

**Comment by romain-rsr** [comment]
Hi,

Thanks a lot for these complementary works on our examples. I focus my experiments on getting the segmentation encoding for a private dataset whom characteristics are halfway between those of the toy dataset and those of highly structured datasets (celeba, flower etc). Since we can't communicate publicly on it, I'll try to find a dataset that share the same problematics (high details and diversity with very low overall structuration)

Bests,
Romain

**Comment by noahzhy** [comment]
What if the number of points is not confirmed? Does it work?
